ports/stm32f3/board_flash.c

```diff
@@ -71,59 +71,76 @@ static bool is_blank(uint32_t addr, uint32_t size)
   return true;
 }
 
-                    //ADDR            data
-void flash_write(uint32_t dst, const uint8_t *src, int len)
+static bool flash_erase(uint32_t addr)
 {
-   // assume sector 0-7 (bootloader) is same size as sector 1
-  uint32_t addr = APP_LOAD_ADDRESS;
+  // starting address from 0x08000000
+  uint32_t sector_addr = FLASH_BASE_ADDR;
+  bool erased = false;
+
   uint32_t sector = 0;
-  int erased = false;
   uint32_t size = 0;
 
-  for ( unsigned i = 0; i < BOARD_FLASH_SECTORS; i++ )
+  for ( uint32_t i = 0; i < SECTOR_COUNT; i++ )
   {
+    TUF2_ASSERT(sector_addr < FLASH_BASE_ADDR + BOARD_FLASH_SIZE);
+
     size = flash_func_sector_size(i);
-    if ( addr + size > dst )
+    if ( sector_addr + size > addr )
     {
-      sector = i + 1;
-      erased = erasedSectors[i];
-      erasedSectors[i] = 1;    // don't erase anymore - we will continue writing here!
+      sector = i;
+      erased = erased_sectors[i];
+      erased_sectors[i] = 1;    // don't erase anymore - we will continue writing here!
       break;
     }
-    addr += size;
+    sector_addr += size;
   }
 
-  if (sector == 0)
-  {
-    TUF2_LOG1("invalid sector\r\n");
-  }
+  (void)sector;
 
-  HAL_FLASH_Unlock();
+#ifndef TINYUF2_SELF_UPDATE
+  // skip erasing sector0 if not self-update
+  TUF2_ASSERT(sector);
+#endif
 
-  if (!erased && !is_blank(addr, size))
+  if ( !erased && !is_blank(sector_addr, size))
   {
-    uint32_t SectorError = 0;
-
-    TUF2_LOG1("Erase: %08lX size = %lu\n", addr, size);
+    TUF2_LOG1("Erase: %08lX size = %lu KB ... ", sector_addr, size / 1024);
 
     FLASH_EraseInitTypeDef EraseInit;
     EraseInit.TypeErase = FLASH_TYPEERASE_PAGES;
-    EraseInit.PageAddress = addr;
-    EraseInit.NbPages = ((0x08040000 - addr)/size);
+    EraseInit.PageAddress = sector_addr;
+    EraseInit.NbPages = ((BOARD_FLASH_APP_START - sector_addr)/size);
 
+    uint32_t SectorError = 0;
     HAL_FLASHEx_Erase(&EraseInit, &SectorError);
     FLASH_WaitForLastOperation(HAL_MAX_DELAY);
-
-    if (SectorError != 0xFFFFFFFF)
-    {
-      TUF2_LOG1("failed to erase!\r\n");
-    }
+    TUF2_LOG1("OK\r\n");
+    TUF2_ASSERT( (SectorError != 0xFFFFFFFF) && is_blank(sector_addr, size) );
```

Comment:
seems to be inverted logic error, it must be `TUF2_ASSERT(SectorError == 0xFFFFFFFF)` since 0xFFFFFFFF all pages is erased.